tiny_dnn/util/product.h

```diff
@@ -516,6 +516,68 @@ void fill(T *dst, size_t size, T value) {
 
 }  // namespace detail
 
+#ifdef CNN_USE_AVX
+#ifdef CNN_USE_DOUBLE
```

Comment:
Please check other implementation in `tiny_dnn/util/product.h` file.
There are several classes.

```cpp
detail::double_avx
detail::float_avx
detail::double_sse
detail::float_sse
detail::scalar_generic<double>
detail::scalar_generic<float>
```

With ugly template metaprogramming, we can reduce code duplications.